README
WHAT IS CGIPAF?

   "cgipaf" allows users to change their un*x Password, Autoreply and
   mail Forwarding by a web interface.
   
REQUIREMENTS
  
    + a Un*x server
    + a mailserver
    + webserver
    + gcc, ndbm or gdbm
    
BUGS
   
    see BUGS
    
LICENSE

    SOFTWARE
   
      CGIpaf (the software) is licensed under the terms of
      the GNU General Public License version 2 or later.

      Copyright (c) 2002 - 2020 Staf Wagemakers.

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 3 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <https://www.gnu.org/licenses/>.

    DOCUMENTATION

      The documentation for CGIpaf is licensed under the terms of the
      FreeBSD Documentation License.

      The FreeBSD Documentation License

      Copyright (c) 2001 - 2020 Staf Wagemakers

      Redistribution and use in source (plaintext, HTML) and 'compiled' forms
      (SGML, HTML, PDF, PostScript, RTF and so forth) with or without
      modification, are permitted provided that the following conditions are
      met:

      1. Redistributions of source code (plaintext, HTML) must retain the
         above copyright notice, this list of conditions and the following
         disclaimer.

      2. Redistributions in compiled form (transformed to other DTDs,
         converted to PDF, PostScript, RTF and other formats) must reproduce
         the above copyright notice, this list of conditions and the following
         disclaimer in the documentation and/or other materials provided with
         the distribution.

      THIS DOCUMENTATION IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
      WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
      MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
      NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY DIRECT,
      INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
      (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
      SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
      HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
      STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
      ANY WAY OUT OF THE USE OF THIS DOCUMENTATION, EVEN IF ADVISED OF
      THE POSSIBILITY OF SUCH DAMAGE. 

INSTALLATION

    see INSTALL

OVERVIEW

    cgipaf is a combination of three cgi programs:
    
    passwd.cgi:       allow users to update their password
	  viewmailcfg.cgi:  allow users to view their current mail 
	                   configuration.
	  mailcfg.cgi:      update the mail configuration
	 
    All programs use PAM for user authentication, systems without PAM
    are only supported if they use the standard password file location
    (/etc/passwd /etc/shadow) and the standard crypt function to encrypt
    the passwords.
    
    If a password is changed it is possible to run a script to update
    SAMBA passwords, NIS configuration, etc.
	 
    The mail configuration is done by procmail by default, mailcfg.cgi 
    creates a .procmailrc in the user's home directory. If autoreply is 
    enabled the autoreply message is stored in ~/vacations.txt. When a user 
    enables mail forwarding the user can choose to keep the messages.
    
    If you want to use another auto responder like vacation you can define
    an external script to update the user's mail configuration. See examples
    for an example config.
    
    For user authentication viewmailcfg.cgi creates a cookie, this cookie 
    is stored in the accessdb. mailcfg.cgi reads the cookie out the accessdb 
    and compares it with the cookie send by the users browser.
    
    User with too many invalid logins can be locked.
    
    The minimum and maximum uid can be set in the configuration file, so
    you can specify a range of uid's that are allowed to use cgipaf.
  
    You can specify html files with a few PHP extensions for the login screen
    and error messages.
    
    It's possible to use a redirect for the error messages, instead of plain 
    html files.

CONFIGURATION
 
    see Configuration.html

CONTACT INFO

    email:  staf@wagemakers.be
    
    Future improvements will be available at 
    http://www.wagemakers.be/english/programs/cgipaf/


ChangeLog
* version 1.3.5		( 13 Apr 2016)
  + BUGFIX: pwd_mkdb errors were not handled correctly on BSD 
  + kyua test scripts added 

* version 1.3.4		( 23 Nov 2014 )
  + Cracklib configuration checking has been improved
  + LDFLAGS is passed to the linker correctly

* version 1.3.4pre1	( 15 Sep 2013 )
  + Enable PAM on FreeBSD 7.3 or above  
  + Enable PAM on NetBSD 6.0 or above
  + xmalloc updated to support systems with non GNU compatible malloc

* version 1.3.3		( 24 Jan 2008 )
  + BUGFIX: md5 passwords on BIG ENDIAN systems were not calculated correctly

* version 1.3.2		( 30 Aug 2007 )
  + changepass improved
  + added formail directive
  + BUGFIX: lock dir is detected correctly
  + BUGFIX: corrected compile warnings
  + BUGFIX: vacation_cgi example has been update to work with newer
            versions of perl/cgi module

* version 1.3.1		( 13 jul 2004 )
  + BUGFIX: install.sh will now ask to create the cgi-bin directory
            (Thanks to Koen de Jong for reporting this problem)
  + BUGFIX: configure --disable-pam works now on PAM systems
  + BUGFIX: fixed compile problem on non-PAM systems
            (Thanks to Peter Hübert for reporting this problem)
  + Dutch translation added  (Thanks to Koen de Jonge)   

* version 1.3.0	        ( 26 Apr 2004 )
  + Documentation updated

* version 1.3.0rc2	( 10 Mar 2004 )
  + PAM: the pam functions should now work on Solaris 9+ 

* version 1.3.0rc1	( 6 Jan 2004 )
  + autoconf has been upgraded to 2.57: the build host is auto-detected
  + new variables:
        _POST[] array : An associative array of variables with the original
	                HTTP POST
        post_string   : A string with the original HTTP POST
	message       : A string with the default message
  + new directives to enable more advanced mail configuration scripts:
    run_viewmailcfg, mailcfg_check. a sample perl vacation script is included.
  + set_script_filename and unset_script_filename directives 
    ( useful if you want to use a php script )

* version 1.3.0pre4    	( 30 May 2003 )
  + new forward method to avoid mail looping

* version 1.3.0pre3    	( 28 Apr 2003 )

  + BUGFIX: setuid(0) added before the passwd run_success script, to
    re-enable smbpasswd support. which was broken in the previous release.

* version 1.3.0pre2    	( 31 Mar 2003 )
  
  + NetBSD support
  + BUGFIX: 1.3.0pre1 didn't compile on FreeBSD
  
  + merged from CGIpaf 1.2.1
   
     - example in Brazilian Portuguese added ( Thanks to Fabricio Machado )
     - improved forward_to emailaddress testing
     - added test for mail headers like 'Precedence: bulk' to avoid replying 
       to maillinglists etc.
     - code cleanup

* version 1.3.0pre1     (  2 Jan 2003 )
  + linuxconf virtual mail password files support added

* version 1.2.0         (  7 Dec 2002)
  + 1.2.0RC1 has been released as 1.2.0 

* version 1.2.0RC1      ( 26 Nov 2002 )
  + setreuid(user_uid,0) added in cgipaf.c 
    passwd.cgi should now support all pam options in the PAM configuration 
    (cracklib, minimum/ maximum password length etc.) on all systems with PAM
    support.

  + The set_PAM_CHANGE_EXPIRED_AUTHTOK directive is obsolete and will be 
    removed in the next releases of CGIpaf

* version 1.2.0pre1     ( 24 Oct 2002 )
  + Access Control List added ( AclOrder, AllowUsers, AllowGroups, DenyUsers
    DenyGroups directives )

* version 1.1.0         ( 17 Sep 2002 )
  + Documentation update

* version 1.1.0rc2      (  07 Aug 2002 )
  + The ability to list words in /etc/cgipasswd that should not be a part of
    new passwords  ( Thanks to Daniel R for his patch )

* version 1.1.0rc2pre1  (  24 Jul 2002 )
  + FIXED: run_after_mailcfg was executed before the mail configuration

* version 1.1.0rc1      (  15 Jul 2002 )
  + changepass utility added

* version 1.1.0pre2      ( 26 Jun 2002 )
  + new run_mailcfg directive added to support other mail configurations than 
    procmail a sample script with vacation support is included.

* version 1.1.0pre1      ( 18 Jun 2002 )
  + new run_before_mailcfg and run_after_mailcfg directive
  + new use_statefile directive

* version 1.0.1          (  4 Jun 2002 )
  + improved cracklib detection

* version 1.0.1rc1       ( 21 May 2002 )
  + updates to configure
  + documentation update

* version 1.0.1rc1pre2   ( 17 May 2002 ) 
  + FreeBSD support

* version 1.0.1rc1pre1   ( 29 Apr 2002 )
  + md5 password support on non-pam systems ( Slackware )

* version 1.0.0          ( 23 Apr 2002 )
  + faq
  + documentation update
  + new install script

* version 1.0.0pre1.2    ( 3 Apr 2002 )
  + more logging in debug mode
  + documentation update

* version 1.0.0pre1.1    ( 28 Mar 2002 )
  + new password unchanged directive
  + updates to pass.c to support more non-GNU/Linux Unices

* version 1.0.0pre1.0    ( 19 Mar 2002 )
  + new cracklib_dictpath directive
  + FIXED cgipaf did't compile on systems where CRACKLIB_DICTPATH wasn't 
    defined
  + update to configure

* version 1.0rc5         ( 28 Jan 2002 )
  + documentation update
  + php files update

* version 1.0rc5pre2     ( 16 Jan 2002 )
  + new pam conversation function
  + new set_PAM_CHANGE_EXPIRED_AUTHOK directive

* version 1.0rc5pre1     ( 20 Dec 2001 )
  + cracklib support
  + support for all common pam messages ( cracklib, minimum password length etc )

* version 1.0rc4     ( 4 Dec 2001 )
  + documentation update

* version 1.0rc4pre5 ( 28 Nov 2001 )
  + new pam_service directive, you can change the pam service name.
  + added test for gdbm/ndbm.h header file
  + fixed: mailcfg.cgi failed, if it couldn't determine the domain name.
  
* version 1.0rc4pre4 ( 11 Nov 2001 )
  + new forwarding method to avoid mail looping
  + fixed: it wasn't possible to use special characters ( $ space ) as the
    loginname / password.

* version 1.0rc4pre3 ( 27 oct 2001)
  + fixed: run_success, run_locked didn't always work due a buffer overflow in
    xstring.c
  + fixed: typo the sample samba passwd script the configuration file & 

pass.c
/*
 * pass.c
 *
 * Copyright (C) 2000,2002,2007,2015,2016 Staf Wagemakers Belgie/Belgium
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "pass.h"
#include <stdio.h>
#include "fgetpwnam.h"

/*
 * if BSDHOST is not defined assume /etc/passwd
 */
#ifndef BSDHOST
#define PASSWDFILE	"/etc/passwd"
#else
#define PASSWDFILE	_PATH_MASTERPASSWD
#endif

#ifdef MD5_CRYPT
#include "md5crypt.h"
#endif

static char *passwd_location=NULL;
static char *shadow_location=NULL;
static int  *sha2_prefered_rounds=NULL;

/*
 * set the prefered rounds for sha2
 */

int * set_sha2_prefered_rounds(int i) {

	if(sha2_prefered_rounds!=NULL) {
		xfree(sha2_prefered_rounds);
	}

	if(i==0) {

		sha2_prefered_rounds=NULL;

	} else {

		sha2_prefered_rounds=xmalloc(sizeof(int));
		*sha2_prefered_rounds=i;
	}

	return(sha2_prefered_rounds);

};

/*
 * returns the prefered_rounds value
 */

int get_sha2_prefered_rounds() {

	if(sha2_prefered_rounds==NULL) return(0);

	return(*sha2_prefered_rounds);


};

/*
 * Set the password file location 
 * NULL = return current location
 */
char * set_passwd_location ( char * passwd_file ) {
	if(passwd_file!=NULL) passwd_location=passwd_file;
	return(passwd_location);
}

/*
 * Set to passwd_location to NULL
 * ( use system passwd )
 */
void reset_passwd_location () {
	passwd_location=NULL;
}

/*
 * Set the shadow file location
 * NULL = return the current location
 */
char * set_shadow_location ( char * shadow_file ) {
	if(shadow_file!=NULL) shadow_location=shadow_file;
	return(shadow_location);
}

/*
 * Set to shadow_location to NULL
 * ( use system shadow file )
 */
void reset_shadow_location() {
	shadow_location=NULL;
}


#ifndef HAVE_SHADOW_H
/*
 * no lckpwdf so create our own
 */
int lckpwdf () 
{
struct stat st;
int lock;
int i=0;
char *passwdfile=PASSWDFILE;
for(;;) {
    if((lock = open(passwdfile,O_RDONLY,0))<0) return(-2);
    if(fcntl(lock,F_SETFD,1)==-1) return(-2);
    while (flock(lock,LOCK_EX|LOCK_NB)) {
          ++i;
          if(i>30) return(-2);
          sleep(1);
    };
    if(fstat(lock,&st) < 0) return(-2);
    if(st.st_nlink !=0 ) break;
    close(lock);
    lock=-2;
}
return(lock);   
} 
#endif

/*
 * reads the passwd info out /etc/passwd and
 * /etc/shadow
 *
 * name = loginname
 * returns         passwd info
 *                 NULL = error
*/
#ifndef _WITHPAM
struct pw_info * get_pw(char *name)
#else
struct pw_info * get_pw_nopam(char *name)
#endif
{
struct pw_info *pw;
FILE *pw_file=NULL,*sh_file=NULL;
pw=(struct pw_info *) xmalloc(sizeof(struct pw_info));
pw->p=NULL;
pw->sp=NULL;

#ifndef BSDHOST
if(passwd_location==NULL) {
#endif
  	if(!(pw->p=getpwnam(name))) return(NULL);   /* User doesn't exist... */
#ifndef BSDHOST
}
else {
	if((pw_file=fopen(passwd_location,"r"))==NULL) return(NULL);
	if(!(pw->p=fgetpwnam(pw_file,name))) return(NULL);
}
#endif

  if (!strcmp(pw->p->pw_passwd,"x")) {

#ifdef HAVE_SHADOW_H

#ifndef BSDHOST
     if(set_shadow_location(NULL)==NULL) {
#endif
       if(!(pw->sp=getspnam(name))) return(NULL);
#ifndef BSDHOST
     }
     else {
	     if((sh_file=fopen(shadow_location,"r"))==NULL) return(NULL);
	     if(!(pw->sp=fgetspnam(sh_file,name))) return(NULL);
     }
#endif


#else
   /*
    * shadow.h is unavailable, unable to copy useful shadow info in pw
    */
   pw->sp=NULL;
#endif
   }
if(pw_file!=NULL) fclose(pw_file);
if(sh_file!=NULL) fclose(sh_file);
return(pw);
}

/*
 * returns a pointer to the password field
 */
char * get_pwfield (struct pw_info *pw) 
{

     if (pw == NULL) return NULL;

     if(pw->sp==NULL) return(pw->p->pw_passwd);
#ifdef HAVE_SHADOW_H
     return(pw->sp->sp_pwdp);
#else
     return(NULL);
#endif
}

/*
 * returns password crypt type
 * 	  -1 = error
 *         0 = std crypt
 *         1 = md5
 *         2 = blowfish
 *         3 = sha256
 *         4 = sha512
 */
int get_crypttype(struct pw_info *pw)
{
char *pass=get_pwfield(pw);
return(hash2crypttype(pass));
}

/*
 * test a passwd
 * *p = passwd info
 * *pass = password
 * returns:        -1 = wrong passwd
 *                 0 = ok
*/
#ifndef _WITHPAM
int ckpw(struct pw_info *pw,char *pass)
#else
int ckpw_nopam(struct pw_info *pw,char *pass)
#endif
{
char *c,*p,salt[3];
if((p=get_pwfield(pw))==NULL) return(-1);  /* error! */
if (is_des(p)) {
   salt[0]=p[0];  /* p in no md5 password use std salt; */
   salt[1]=p[1];
   salt[2]='\0';
   c=salt;
} else {
   c=p;  	  /* p is a md5 password or better use complete password as salt */ 
}
if (strcmp(p,crypt(pass,c))) return(-1);   /* wrong password */
return(PASS_SUCCESS);
}

char * str_passerror(int passerror) {

 	if(passerror > 0) return("Error is higher than 0 should be ok");

	passerror=abs(passerror);
	
	char * passErrorMessages[] = {
		"ok",					/* 0 	= ok 					*/
		"error open pwfilename", 		/* -1  	= error open pwfilename 		*/
          	"locking failed", 			/* -2  	= locking 				*/
		"locking failed", 			/* -3  	= locking 				*/
 		"error open TMPFILE", 			/* -4  	= error open TMPFILE 			*/
 		"fileno failed", 			/* -5  	= fileno failed				*/
 		"fchmod failed", 			/* -6  	= fchmod failed				*/
 		"bufferlength too small", 		/* -7  	= bufferlength too small		*/
 		"error updating tmpfile", 		/* -8  	= error updating tmpfile		*/
 		"out of memory", 			/* -9  	= out of memory				*/
 		"lsstat failed", 			/* -10 	= lsstat failed				*/
 		"rename failed", 			/* -11 	= rename failed				*/
 		"user not found", 			/* -12 	= user not found			*/
 		"unsupported crypt type",		/* -13 	= unsupported crypt type		*/
 		"pw_mkdb failed ( bsd only )", 		/* -14 	= pw_mkdb failed ( bsd only 	)	*/
 		"encrypt_pass is NULL",  	   	/* -15 = encrypt_pass is NULL 			*/
 		"crypt_newhash failed (OpenBSD only)", 	/* -16 = crypt_newhash failed (OpenBSD only)    */
 		"xcrypt failed (NULL)",  		/* -17 = xcrypt retruns NULL 			*/
 		"unlink oshadow failed",  		/* -18 = unlink oshadow failed 			*/
 		"ferror on reading pwfile",  		/* -19 = ferror on reading pwfile 		*/
		};

	int max=sizeof(passErrorMessages)/sizeof(char *);

	if(passerror > max) return("unkown error");
	return(passErrorMessages[passerror]);
}

/*
 * update the password file
 * *pwfilename = passwd filename
 * *name = username
 * *encrypt_pass = new encrypted password
 * returns  0  = ok
 *         -1  = error open pwfilename
 *         -2  = reserved (locking)
 *         -3  = reserved (locking)
 *         -4  = error open TMPFILE
 *         -5  = fileno failed
 *         -6  = fchmod failed
 *         -7  = bufferlength too small
 *         -8  = error updating tmpfile
 *         -9  = out of memory
 *         -10 = lsstat failed
 *         -11 = rename failed
 *	   -12 = user not found
 *         -13 = reserved (unsupported crypt type)
 * 	   -14 = pw_mkdb failed ( freebsd only )
 * 	   -15 = encrypt_pass is NULL
 */
int update_pwfile(char *pwfilename,struct pw_info *pw,char *encrypt_pass)
{

FILE *pwfile;
FILE *tmpfile;
int fd=0;
int uf=0;                 /* user found, set if user is found */
char buffer[BUFFERLEN];
char buffer2[BUFFERLEN];
char *cp;

struct stat st;
char *name=pw->p->pw_name;

umask(0377);

if (!(pwfile=fopen(pwfilename,"r"))) return(-1); /* can't open password file */
if (!(tmpfile=fopen(TMPFILE,"w"))) return(-4);   /* can't create tmp file    */
if ((fd=fileno(tmpfile))==-1) return(-5);        /* can't convert stream to int file */
if (fchmod(fd,256)==-1) return(-6);              /* can't chmod tmpfile */
if (encrypt_pass==NULL) return(-15);		 /* encrypt_pass is NULL */

/*
 * password file parser
 */
   
while (fgets(buffer,BUFFERLEN-1,pwfile)) { 
   if(!(strrchr(buffer,'\n'))) {errno=E2BIG;return(-7);} /* buffer overflow */
   if(!uf) {                     
   if(!(*buffer=='#' || *buffer=='\n')) {     /* if buffer is no comment or empty */
   cp=(char *)strchr(buffer,':');  /* cp point to the end of the current username */
   if(cp==NULL) return (-2);      /* ":" not found, doesn't look like a password file... */
   strncpy(buffer2,buffer,cp-buffer);
   buffer2[cp-buffer]='\0';        /* buffer2 = current username */
   if (strcmp(name,buffer2)==0) {
      uf=1;                        /* found the user ready to update his password */
      strcpy(buffer2,buffer);
      cp=(char *)strchr(buffer2,':');
      cp=(char *)strchr(cp+1,':'); /* cp point to the end of password field */
      if ((strlen(cp)+strlen(name)+strlen(encrypt_pass))>BUFFERLEN) {
         errno=E2BIG;
         return(-7);               /* buffer overflow */
	 }
      strcpy(buffer,name);
      
      /*
       * create new passwd line
       */
      
      strcat(buffer,":");
      strcat(buffer,encrypt_pass);
      strcat(buffer,cp);
   }
   }
   }
   /*
    * copy the buffer to the temp file
    */
   
   if (fputs(buffer,tmpfile) == EOF) return(-8);
} 

/*
 * return error on ferror
 */

if (ferror(pwfile)) return(-19);

fclose(tmpfile);
fclose(pwfile);
/*
 * update to permissions of tmpfile to the permissions of pwfilename
 */
if (lstat(pwfilename,&st)==-1) return(-10);
chmod(TMPFILE,st.st_mode);
chown(TMPFILE,st.st_uid,st.st_gid);

if (uf) {
   /* user found, rename tmpfile to pwfile */
#ifndef BSDHOST
   if (rename(TMPFILE,pwfilename)==-1) return(-11); 
#else

/*
 * Free|Net Bsd password changer
 * use pw_mkdb (8) to update the passwd db, TMPFILE is the input file
 */
   int pstat;
   pid_t pid;

#ifdef NETBSDHOST
/* use / if NETBSD */
   char *mppath = "/";
   char *pwd_mkdb_file=TMPFILE;
#else
#ifdef OPENBSDHOST
/* use /etc if OpenBSD */
   char *mppath = "/etc";
   char *pwd_mkdb_file=TMPFILE;
   /* assume FREEBSD */
#else
   char *mppath = _PATH_PWD;
   char *pwd_mkdb_file=TMPFILE;
#endif
#endif
   
   if(!( pid=fork() )) {

      	if (!name) {

         	execl(_PATH_PWD_MKDB,"pwd_mkdb","-p","-d",mppath,pwd_mkdb_file,NULL);

	}
      	else {

         	execl(_PATH_PWD_MKDB,"pwd_mkdb","-p","-d",mppath,"-u",name,pwd_mkdb_file,NULL);

	}

      return(-14);
   }
   pid = waitpid(pid,&pstat,0);

   if ((pid == -1 || !WIFEXITED(pstat) || WEXITSTATUS(pstat) != 0)) return(-14);
#endif
}
else { 
   /* user not found, nothing to do */
   errno=0; 
   return(-12);
}
errno=0;
return(PASS_SUCCESS);
}

#ifndef _WITHPAM
int chpw(struct pw_info *pw,char *pass)
{

	int ret;
	ret=chpw_nopam(pw,pass,255);

	return(ret);

}

#endif

/*
 * update the user's password
 *
 * mode:	
 * 		0:  force des passwors
 * 		1:  force md5 passwords
 * 		2:  force blowfish
 *		3:  force sha256 passwords
 *		4:  force sha512 passwords 
 *		254:  already encrypted
 * 		255:  use existing crypttype
 *
 * 
 * returns:
 *             -2:  lockfile exists
 *             -3:  can't create lockfile
 *             -13:  unsupported crypt type
 */

int chpw_nopam (struct pw_info *pw, char *pass,int mode)
{
int i=0;
int fd_tmplock=0;
int lock=0;
int count=0;
char *passwdfile=PASSWDFILE;
char *encrypt_pass=NULL;
struct stat st;

if (mode<5) {

	if(!is_crypt_id_supported(mode)) {

		return(-13);


	}

}

/*
 * wait until the lockfile is unlocked
 */
count=10;
while (count--) {
   if ((lock=lckpwdf())!=-2) break;
   sleep(3);
   } 

/*
 * lockfile exists and we've waited long enough
 */
   
if (lock==-1) return(-2);

/* 
 * create lockfile
 */
   
count=10;
while (count--) {
   if ((fd_tmplock=open(TMPLOCK, O_RDWR | O_CREAT | O_EXCL, 0444))==-1) {
      sleep(3);
      }
      else { 
         close(fd_tmplock);
	 break;
	 }
      }

/*
 * can't create lockfile
 */
   
if (fd_tmplock==-1) return(-3); /* can't create lock */

switch(mode) {

	case 0:
		/*
 		 * force DES passwords if mode=0 	
 		 */
		i=0;
		break;

	case 1:

		/*
 		 * force md5 passwords if mode=1
 		 */

		i=1;
		break;
	case 2:
		i=2;
		break;

	case 3:
 
		/*
 		 * force sha256 passwords if mode=3
 		 */

		i=3;
		break;

	case 4:
 
		/*
 		 * force sha512 passwords if mode=3
 		 */

		i=4;
		break;

	case 10:

		/*
		 * force sha1 password if mode=10
		 */

		i=10;
		break;

	case 254:
		encrypt_pass=pass;
		i=254;
		break;

	default:
		i=get_crypttype(pw);

		if( (i==-1) || !is_crypt_id_supported(i) ) {

				i=xcrypt_best_supported_crypt_id();

		}

}

/*
 * Create an encrypted password
 */


if (mode !=254 ) {


	switch(i) {

		case 0:
		case 1:
		case 2:
		case 10:
              		encrypt_pass=xcryptbyid(pass,i,NULL);
	     		break;
		case 3:
		case 4:
              		encrypt_pass=xcryptbyid(pass,i,sha2_prefered_rounds);
	     		break;
       		default:
			encrypt_pass=NULL; /* unsupported crypt type! */
	       		i=-13;

	}

}

if (encrypt_pass!=NULL) {

	/*
 	* set passwdfile to the real password file
 	*/

	if (pw->sp) { 

		if((passwdfile=shadow_location)==NULL)
		passwdfile=SHADOWFILE;

   	}
   	else {

	   	if((passwdfile=passwd_location)==NULL) 
		   passwdfile=PASSWDFILE;

   	}

	/*
 	* try to update the user's password
 	*/

	i=update_pwfile(passwdfile,pw,encrypt_pass);

} else {

	i=-17;

}

unlink(TMPLOCK);

#ifndef BSDHOST
ulckpwdf();
#endif

if (i<0) return(i);   /* can't update password */

/*
 * Solaris compatibility
 */

#ifdef SOLARISHOST


if (!strcmp(passwdfile,SHADOWFILE)) {
   if(lstat(OSHADOWFILE,&st)!=-1) {
      if(unlink(OSHADOWFILE)!=0) return(-18);
      }
   }
#endif

return(PASS_SUCCESS);
}

